pkg/pma/pma.c

```diff
@@ -0,0 +1,860 @@
+static inline page_status_t
+page_status_(const void *addr, const pma_t *pma)
+{
+    size_t  pg_idx    = addr_to_page_idx_(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t status    = (pma->pg_status[entry_idx] >> bit_idx) & PS_MASK;
+    return status;
+}
+
+/// Set the status of the page surrounding an address. To get rather than set,
+/// see page_status_().
+///
+/// @param[in] addr    Address within the page in question. Must be within the
+///                    bounds of the PMA.
+/// @param[in] status  New status of the page in question.
+/// @param[in] pma     PMA the page in question belongs to.
+static inline void
+set_page_status_(const void *addr, page_status_t status, const pma_t *pma)
+{
+    size_t  pg_idx    = addr_to_page_idx_(addr, pma);
+    size_t  entry_idx = pg_idx / kPagesPerEntry;
+    size_t  bit_idx   = (pg_idx % kPagesPerEntry) * kBitsPerPage;
+    uint8_t entry     = pma->pg_status[entry_idx];
+    pma->pg_status[entry_idx]
+        = (entry & ~(PS_MASK << bit_idx)) | (status << bit_idx);
+}
+
+/// Set the status of the page range starting at an address. To set the page
+/// status of a single page, see set_page_status_().
+///
+/// @param[in] addr    Address within the first page in question. Must be within
+///                    the bounds of the PMA.
+/// @param[in] pg_cnt  Number of pages in the range. The entirety of the page
+///                    range must be within the bounds of the PMA.
+/// @param[in] status  New status of the page range in question.
+/// @param[in] pma     PMA the page range in question belongs to.
+static inline void
+set_page_status_range_(const void   *addr,
+                       size_t        pg_cnt,
+                       page_status_t status,
+                       const pma_t  *pma)
+{
+    for (size_t i = 0; i < pg_cnt * kPageSz; i += kPageSz) {
+        set_page_status_((char *)addr + i, status, pma);
+    }
+}
+
+/// Sync changes made to either the heap or stack since the last call to
+/// pma_sync().
+///
+/// @param[in] path        Path to backing file. If NULL, this function is a
+///                        no-op.
+/// @param[in] base        Base address of the file-backed mapping.
+/// @param[in] grows_down  true if the mapping grows downward in memory.
+/// @param[in] pma         PMA this mapping belongs to.
+/// @param[in] len         Number of bytes at the beginning of the mapping to be
+///                        synced to the backing file. Must be a multiple of
+///                        kPageSz.
+/// @param[in] fd          Open file descriptor of the backing file.
+///
+/// @return 0   Synced changes successfully.
+/// @return -1  Failed to sync changes.
+static int
+sync_file_(const char *path,
+           void       *base,
+           bool        grows_down,
+           pma_t      *pma,
+           size_t      len,
+           int         fd);
```

Comment:
I think this api would look more consistent if it accepted two file descriptors rather than a path for the wal and an fd for the persistent loom image (home heap/stack).

Additionally wouldn't have to open a new file descriptor each time we write the wal.

And at the very least, I think `path` needs to be renamed and its comment updated. i.e. call out that this is the path _to the wal_ (not the other backing file).

Reply:
> path is the path to the backing file, which is used as the basename of the WAL

Oh I see. I was confused because path is only passed to the function to create the wal file. It might be more clear if path and fd were wrapped in a struct since that data is tied together. Perhaps the definition of `struct pma` could be slightly modified to this end. You may even add more members to the backing file struct like `len` to replace [stack|heap]_len Like:

```c
typedef struct backing_file {
  const char *path;
  int         fd;
} backing_file_t

static int
sync_file_(backing_file_t *f,
           const char *path,
           void       *base,
           bool        grows_down,
           pma_t      *pma,
           size_t      len);

/// Persistent memory arena handle.
typedef struct pma {
    backing_file_t heap;
    backing_file_t stack;

    /// Base address of the heap.
    void *heap_start;

    /// Base address of the stack. Points to one byte past the end of the PMA.
    void *stack_start;

    /// Number of bytes of the heap mapped into memory. Guaranteed to be a
    /// multiple of kPageSz.
    size_t heap_len;

    /// Number of bytes of the stack mapped into memory. Guaranteed to be a
    /// multiple of kPageSz.
    size_t stack_len;

    /// Total number of pages in the PMA. Each page is `kPageSz` bytes.
    size_t num_pgs;

    /// Bit map for tracking pages, which can be in one of three states:
    /// unmapped, mapped and clean, or mapped and dirty (see page_status_t).
    /// Guaranteed to have at least (2 * num_pgs) bits.
    uint8_t *pg_status;

    /// Function used to get active heap and stack lengths.
    len_getter_t len_getter;

    /// Base address of the guard page.
    void *guard_pg;

    /// Out-of-memory handler.
    oom_handler_t oom_handler;

    size_t        max_sz;

    /// libsigsegv ticket.
    void *sigsegv_ticket;
} pma_t;
```